### PR TITLE
feat: add `list_price` to response of can_redeem

### PR DESCRIPTION
**Pull request by pwnage101**
This also bundles other improvements:
* output serializers are defined for the can_redeem endpoint, enabling much higher quality redoc pages
* added a ContentKeyField to replace the ValidateContentKeyMixin which was harder to use and overly complicated.

ENT-7137

### CODE REVIEW

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (pwnage101) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:610]: FYI, the code as written just passes-through a string representing decimal dollars (which is how the list_price is stored in the content metadata upstream), rather than integer cents used in the rest of this service, including in other API responses.  My personal opinion would be to have this convert the output to be consistent with the rest of the quantities returned by this API, but I also was just referencing the API spec: https://github.com/openedx/enterprise-access/blob/6ad37aac9b3fbdf594c2409fd8b7ad8aa7ab3555/docs/decisions/0006-api-specification-for-enterprise-mfes.rst#L1

Open to suggestions!

Comment (pwnage101) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:610]: FYI, I just looked at the API spec and discovered that integer cents was called for, not a passthrough string representing decimal dollars. https://github.com/openedx/enterprise-access/blob/6ad37aac9b3fbdf594c2409fd8b7ad8aa7ab3555/docs/decisions/0006-api-specification-for-enterprise-mfes.rst#L107

I plan to convert this before merging.

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (iloveagent57) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:631]: Nice, way easier to reason about everything wrapped up in here now.

**Review by iloveagent57 — APPROVED** [review]
lgtm, I'll take another look once you do that stuff with `list_price`.

**Comment by pwnage101** [comment]
see 2nd commit for changes.  changed the list_price output to an integer, and also refactored the way of mocking the API to get around the lru_cache persisting between tests.

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (adamstankiewicz) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:631]: This is technically a breaking change to update the response from a list to a dict with a `results` list, as the client (learner portal MFE) currently expects a list to be returned. If we change this, it will likely result in a JavaScript error when parsing its response in the learner portal course page on stage.

A couple links to learner portal source where we parse the `can-redeem` response:
* https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/2e8ebf2beac823651019338352dd2841d7b860c2/src/components/course/data/hooks.jsx#L552-L557
* https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/2e8ebf2beac823651019338352dd2841d7b860c2/src/components/course/data/hooks.jsx#L567

image

A few things to consider: 
1. be OK with potentially breaking stage learner portal course page until it is updated to reflect this change shortly after (will not impact production).
2. version the endpoint.
3. update learner portal course page to handle both style of responses and then remove the old way.
4. something else?

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (adamstankiewicz) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:612]: [curious] Is it worthwhile to return both the integer cents and the decimal dollars? It feels a bit odd for the API to convert the list price to cents only for the caller of the API (learner portal) to convert it back to decimal dollars. A potential idea:

```js
{
  "list_price": {
    "USD": 299,
    "USD_CENTS": 29900,
  }
}
```

File: enterprise_access/apps/api/serializers/subsidy_access_policy.py
Comment (adamstankiewicz) [reply enterprise_access/apps/api/serializers/subsidy_access_policy.py:99]: +1 this is a nice addition :)

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (adamstankiewicz) [reply enterprise_access/apps/api/v1/views/subsidy_access_policy.py:612]: Related, locally, I'm seeing a course price of $299.00 in course-discovery be returned as `299000` in the `can-redeem` response, which would equate to `$29,900`.

It looks like `get_subsidy_content_data` is already returning integer cents?

Comment (adamstankiewicz) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:612]: If `list_price_decimal_dollars_str` can be `None`, won't an exception be raised here: `float(list_price_decimal_dollars_str)`?

```python
>>> float(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: float() argument must be a string or a number, not 'NoneType'
```

The try/except doesn't catch `TypeError` though, so I believe this exception would be raised.

Comment (pwnage101) [reply enterprise_access/apps/api/v1/views/subsidy_access_policy.py]: handled case, and added new test.

Comment (pwnage101) [reply enterprise_access/apps/api/v1/views/subsidy_access_policy.py]: I think I would agree about being explicit about the type, since the unit in this case doesn't directly pertain to any unit of any particular ledger/subsidy.

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (pwnage101) [reply enterprise_access/apps/api/v1/views/subsidy_access_policy.py]: Good catch, I forgot to mention this.  I re-worked the endpoint to revert it back to original behavior, with ample code comments to explain why we chose to deviate from the industry norm.  Also, I re-worked it in a way that the redoc page looks accurate:
image

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (pwnage101) [reply enterprise_access/apps/api/v1/views/subsidy_access_policy.py]: updated to match your proposed format.  Note that `usd` is a float, and `usd_cents` is an integer.

**Review by adamstankiewicz — APPROVED** [review]
LGTM!

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (iloveagent57) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:257]: FWIW, I think `lru_cache` will cache for the lifetime of the process, not the request lifetime.

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (iloveagent57) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:640]: This method body is getting really long - is there an obvious place to wrap this in a helper method or something?  Or maybe the serializer can do the translation of string to numbers?

**Review by iloveagent57 — APPROVED** [review]
Looks good, two small notes.

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (pwnage101) [reply enterprise_access/apps/api/v1/views/subsidy_access_policy.py:640]: I'd love to consolidate this into the serializer, but in light of my long todo list I'm going to defer this and add a comment.

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py
Comment (pwnage101) [reply enterprise_access/apps/api/v1/views/subsidy_access_policy.py:257]: Interesting, that does seem to match up with my experience with these having lasting memory between unit tests, which in theory should destroy and create new instances of these views and models.  I'll update all the docstrings for now to say that they're process-cached since I think there's enough of a combination of sentiment and evidence to support the claim.  Note that `@cached_property` will have the same behavior since it's effectively backed by `lru_cache`.

**Pull request merged**
